opengovsg/postmangovsg
# feat(frontend): set up test fixtures and deployment processes

## Problem

We want to start testing the frontend.

Closes issue #1078

## Solution

**Features**:

- Mock the Canvas API for Lottie
- Install a shim for the WebCrypto API since jsdom doesn't implement it
- Start listening on the `mswjs` server before all tests
- Use fake Jest timers
- Mock the `muted` property for `HTMLMediaElement` to workaround a React issue
- Pass in optional parameters to configure `react-ga`
- Add script to run all frontend tests during deployment on Amplify and Travis

## Before & After Screenshots

There are no visual changes.

## Tests

The tests have been verified locally, but not on staging yet.

1. Deploy the app on staging.
2. Verify that the automated tests run without errors on Travis and Amplify.
3. Conduct smoke tests to ensure that overall functionality is unaffected.

- [x] Write a simple unit test for rendering the landing page when unauthenticated

## Deploy Notes

The aforementioned tests have been added and should be run during every deployment on Travis and Amplify. If any test fails, the deployment should not go through.

Specifically, Amplify should not publish the new build if any test fails.

**New scripts**:

- `test`: Runs the frontend tests. If `CI=true` is not set, the script will be interactive and auto-reloaded. Otherwise, the script will run all tests exactly once and exit. (ref: https://create-react-app.dev/docs/running-tests/#continuous-integration)

Note: Travis runs with `CI=true` by default, but Amplify doesn't.

**New dev dependencies**:

- `msw`: to mock backend API responses
- `@testing-library/react`: for idiomatic react testing APIs
- `@testing-library/jest-dom`: for idiomatic DOM assertion APIs
- `@testing-library/user-event`: for idiomatic DOM user event APIs
- `jest-canvas-mock`: to mock Canvas for Lottie as jsdom doesn't implement it
- `@peculiar/webcrypto`: as jsdom doesn't implement the WebCrypto API
